# **Introdução**

Em programação, uma função é uma sequência nomeada de instruções que executa uma operação de computador. Ao definir uma função, especifica-se: i) o nome; e a ii) sequência de instruções. Depois disso, é possível chamá-la pelo nome quantas vezes forem necessárias.

## **Chamada de Função**

Já vimos até aqui diversos exemplos de chamadas de [funções embutidas](https://docs.python.org/pt-br/3/library/functions.html#built-in-functions), como por exemplo a função [*input*](https://docs.python.org/pt-br/3/library/functions.html#input):

In [1]:
nome = input('Digite seu nome: ')

Digite seu nome: Tiago


Quando fazemos isso, o terminal ou prompt de comando irá:

- Mostrar o texto "Digite seu nome: " na tela; e
- Aguardar até que o usuário digite alguma informação.

A função para imprimir dados em Python é a função [*print*](https://docs.python.org/pt-br/3/library/functions.html#print). Ela é responsável por mostrar valores em seu terminal:



In [2]:
print('Olá', nome)

Olá Tiago


Em qualquer função, a expressão entre parênteses é chamada de argumento. É comum dizer que uma função recebe um argumento e retorna um resultado. O resultado também é chamado de valor de retorno. Para a primeira função, o resultado é uma string com o que o usuário digitou. Embora a segunda função imprima dados na tela, a mesma não retorna um valor, ou melhor, retorna **None**. Observe abaixo:

In [3]:
retorno = print('Olá', nome)
print(retorno)

Olá Tiago
None


**Atenção:** o objeto Python None denota falta de valor. 

In [4]:
type(retorno)

NoneType

## **Parâmetros da função**

São variáveis declaradas no cabeçalho da função, com uso exclusivo dentro do bloco de instrução da mesma. A definição dos parâmetros que uma função deve receber obriga o envio dos valores a parâmetros sem valores *default*, do contrário, a função não é invocada.

### **O parâmetro `sep` da função *print***

Por padrão, quando utilizamos virgula para separar os itens, a função *print* utiliza espaços para separar cada saída. Porém, podemos utilizar o parâmetro *sep* para definir um caractere de separação. Entenda no exemplo a seguir:

In [5]:
print('Dia', 'Mês', 'Ano')

Dia Mês Ano


In [6]:
print('Dia', 'Mês', 'Ano', sep='/')
print('ontem', 'Hoje', 'Amanhã', sep='-')
print("B", "n", "n", ".", sep='a')

Dia/Mês/Ano
ontem-Hoje-Amanhã
Banana.


### **O parâmetro `end` da função *print***

Por padrão, a função *print* utiliza a quebra de linha '\n' como último caracter. O parâmetro end é responsável por alterar esse comportamento, possibilitando ao desenvolvedor trocar qual caracter será adicionado ao final do dado impresso no terminal. Observe:

In [7]:
# Exemplo com fim de linha sem nenhum caracter
print('Vamos estudar ', end='')
print('Python')

# Exemplo com fim de linha igual à ->
print('Quando ensino', end=' -> ')
print('Aprendo')

# Exemplo com fim de linha igual à :
print("Nota", end=': ')
print(10)

Vamos estudar Python
Quando ensino -> Aprendo
Nota: 10


### **Parâmetro `file` da função *print***

A função *print* também funciona para gravar dados em arquivos. Para isso, utilizamos o parâmetro *file* da função print. Para tanto, precisamos de um arquivo aberto, o que é feito utilizando-se a função open. Veja o exemplo abaixo:

In [8]:
with open('arquivo.txt', 'w') as arquivo:
    print('Primeira3 linha do arquivo', file=arquivo)
    print('Segunda3 linha do arquivo', file=arquivo)

- Para abrir um arquivo, o *Python* possui a função *open*. Ela recebe dois parâmetros: o primeiro é o nome do arquivo a ser aberto, e o segundo parâmetro é o modo que queremos trabalhar com esse arquivo - se queremos ler ou escrever. O modo é passado através de uma string: 'w' (abreviação para write) para escrita e 'r' (abreviação para read) para leitura, 'a' (abreviação para appending) para anexar.

- A instrução *with* é utilizada para garantir a finalização de recursos adquiridos. Contudo, também é possível utilizar utilizar as instruções *try* e *finally*, conforme exemplo abaixo, para abrir o arquivo, escrever e encerrá-lo. 



In [9]:
try:
  arquivo = open('arquivo.txt', 'a')
  print("Terceira linha do arquivo", file=arquivo)
  print("Quarta linha do arquivo", file=arquivo)
finally:
  arquivo.close()

# **Funções Matemáticas**

*Python* possui diferentes módulos (ou bibliotecas) que fornecem diversas funções úteis. Um desses módulos é o *math* que fornece várias funções matemáticas prontas e que são úteis para realizar cálculos matemáticos. Para usar qualquer módulo em *Python* é necessário importá-lo antes. Por exemplo, para o módulo *math* é necessário acrescentar o seguinte comando antes de executar suas funções:

In [10]:
import math

Esta instrução cria um objeto de módulo chamado [*math*](https://docs.python.org/pt-br/3/library/math.html). Ao se exibir o objeto de módulo, são apresentadas informações sobre ele:



In [11]:
math

<module 'math' (built-in)>

A importação é necessária apenas uma vez para cada programa. Após o comando de importação do módulo, é possível usar as funções definidas por ele. Para acessá-las é preciso especificar o nome do módulo e o nome da função, separados por um ponto. Este formato é chamado de notação de ponto. A seguir exploraremos várias funções elementares.

### **math.sqrt(x)**

É uma função para calcular a raiz quadrada de do argumento x.

In [12]:
resultado = math.sqrt(9)
print(resultado)

3.0


### **math.cos(x)**

Função para calcular o cosseno do argumento x, em radianos.

In [13]:
resultado = math.cos(30)
print(resultado)

0.15425144988758405


### **math.tan(x)**

Função para calcular a tangente do argumento x, em radianos.

In [14]:
resultado = math.tan(3.1417)
print(resultado)

0.00010734641061924728


### **math.sin(x)**

Função para calcular o seno de do argumento x.

In [15]:
resultado = math.sin(3.1417)
print(resultado)

-0.00010734641000075741


### **Exercício**

Calcule o fatorial de 10. Dica:  pesquisar na [documentação](https://docs.python.org/pt-br/3/library/math.html).

In [16]:
math.factorial(10)

3628800

### **Principais Funções da Matemática**

Função | Nome | Exemplo | Resultado
---|---|---|---
Raiz quadrada | sqrt | sqrt(9) | 3
Valor absoluto | abs | abs(-2) | 2
Cosseno | cos | cos(30) | 0.1542514
Tangente | tan | tan(7.3456) | 1.7945721
Seno | sin | sin(pi) | 1.2246e-16
Parte inteira | trunc | trunc(5.2) | 5
Maior inteiro menor que ... | ceil | ceil(5.2) | 6
Menor inteiro maior que | floor | floor(5.2) | 5
Potência | pow | pow(2,3) | 8
Logarítmo natural | log | log(148.41315910257657) | 5
Logarítmo base 10 | log10 | log10(1000) | 3
Logarítmo base 2 | log2 | log2(256) | 8
Logarítmo qualquer base | log | log(81,3) | 4

Existe uma variante da instrução import que importa nomes de um módulo diretamente.  Por exemplo:

In [17]:
from math import sqrt
print(sqrt(9))

3.0


### **Constantes**

Além das funções, *Python* possui algumas constantes úteis em diversos cálculos matemáticos fornecidos no módulo math.

#### **pi**

A constante matemática $ $ é utilizada em diversas situações. Uma delas é para se calcular o comprimento de uma circunferência:

In [18]:
from math import pi

r = 100
C = 2 * pi * r
print(C)

628.3185307179587


#### **Número de Euler (ou neperiano)**

Também é muito útil em diversas situações, como em logaritmos naturais por exemplo. Uma outra utilização é para o cálculo de juros compostos, onde o montante após um tempo $t$,$M_t$, a uma taxa de juros composto $r$ é definido como: $M_t=M_i*e^{r*t}$, onde $M_t$ é o montante inicial. O programa a seguir pode ser utilizado para calcular o valor a ser pago considerando um montante
inicial de R$ 1.000,00, com juros composto a uma taxa mensal de 1,5%, durante 12 meses.

In [19]:
from math import e

Mi = 1000
r = 0.015
t = 12
Mt = Mi * e ** (r*t)
print(Mt)

1197.2173631218102


#### **Resumo das constantes**

Constante | Valor
---|---
pi | 3.141592
e | 2.718281
tau | 6.283185
inf | infinito positivo

# **Como Acrescentar Novas Funções?**

Até agora usamos funções que vêm do *Python*, mas também é possível acrescentar novas funções. Uma definição de função especifica o nome de uma nova função e a sequência de instruções que são executadas quando a função é chamada. Abaixo está um exemplo:

In [20]:
def print_refrao():
  print('\nSalve, ó terra dos altos coqueiros')
  print('De belezas soberbo estendal')
  print('Nova Roma de bravos guerreiros')
  print('Pernambuco imortal, imortal!\n')

**def** é uma palavra-chave que indica uma definição de função. O nome da função é print_refrao. As regras para os nomes da função são as mesmas que as das variáveis: letras, números e sublinhados são legais, mas o primeiro caractere não pode ser um número.

**Atenção:** não podemos usar palavras-chave como nome de uma função e devemos evitar ter uma variável e uma função com mesmo nome.

Os parênteses vazios depois do nome indicam que esta função não usa argumentos. A primeira linha da definição de função chama-se cabeçalho; o resto é chamado de corpo. O cabeçalho precisa terminar em dois pontos e o corpo precisa estar endentado, ou seja, em um bloco de instruções. Por convenção, a endentação sempre é de quatro espaços. O corpo pode conter mais de uma instrução.

**Atenção:** bloco de instrução é o conjunto das instrução que estejam num mesmo nível de indentação, mesmo nível hierárquico.

A definição de uma função cria um objeto de função, que tem o tipo function:

In [21]:
print(print_refrao)

<function print_refrao at 0x7f1d8718be60>


A sintaxe para chamar a nova função é a mesma que a das funções integradas:

In [22]:
print_refrao()


Salve, ó terra dos altos coqueiros
De belezas soberbo estendal
Nova Roma de bravos guerreiros
Pernambuco imortal, imortal!



Uma vez que a função tenha sido definida, é possível usá-la dentro de outra função. Por exemplo, para repetir o refrao anterior, podemos escrever uma função print_hino_de_pernambuco:

In [23]:
def print_hino_de_pernambuco():

  print('Coração do Brasil')
  print('Em teu seio corre o sangue de heróis, rubro veio')
  print('Que há de sempre o valor traduzir')
  print('És a fonte da vida e da história')
  print('Desse povo coberto de glória')
  print('O primeiro, talvez, do porvir')
  
  print_refrao()
  
  print('Esses montes, e vales, e rios')
  print('Proclamando o valor de teus brios')
  print('Reproduzem batalhas cruéis')
  print('No presente, és a guarda avançada')
  print('Sentinela indormida e sagrada')
  print('Que defende da pátria os lauréis')
  
  print_refrao()
  
  print('...')
  
  print_refrao()

In [24]:
print_hino_de_pernambuco()

Coração do Brasil
Em teu seio corre o sangue de heróis, rubro veio
Que há de sempre o valor traduzir
És a fonte da vida e da história
Desse povo coberto de glória
O primeiro, talvez, do porvir

Salve, ó terra dos altos coqueiros
De belezas soberbo estendal
Nova Roma de bravos guerreiros
Pernambuco imortal, imortal!

Esses montes, e vales, e rios
Proclamando o valor de teus brios
Reproduzem batalhas cruéis
No presente, és a guarda avançada
Sentinela indormida e sagrada
Que defende da pátria os lauréis

Salve, ó terra dos altos coqueiros
De belezas soberbo estendal
Nova Roma de bravos guerreiros
Pernambuco imortal, imortal!

...

Salve, ó terra dos altos coqueiros
De belezas soberbo estendal
Nova Roma de bravos guerreiros
Pernambuco imortal, imortal!



**Observação:** é preciso criar uma função antes de executá-la, ou seja, a definição de função tem que ser executada antes que ela seja chamada.

## **Fluxo de Execução**

Para garantir que uma função seja definida antes do seu primeiro uso, é preciso saber a ordem na qual as instruções serão executadas. A execução sempre inicia na primeira instrução do programa. As instruções são executadas uma após a outra, de cima para baixo.

As definições de função não alteram o fluxo de execução do programa, mas lembre-se de que as instruções dentro da função não são executadas até que a função seja chamada. Uma chamada de função é um desvio no fluxo de execução. Em vez de ir para a instrução seguinte, o fluxo salta para o corpo da função, executa as instruções lá, e então volta para continuar de onde parou.

## **Parâmetros e Argumentos**

Algumas funções que vimos até aqui exigem argumentos. Por exemplo, ao chamar *math.sin*, você usa um número como argumento. Algumas funções exigem mais de um argumento: o *math.pow* exige dois, a base e o expoente. Dentro da função, os argumentos são atribuídos a variáveis chamada **parâmetros**. Abaixo é definida uma função que precisa de um argumento:

In [25]:
def print_2_vezes(texto):
  print(texto)
  print(texto)

Esta função atribui o argumento a um parâmetro chamado texto. Quando a função é chamada, ela exibe o valor do parâmetro duas vezes.

In [26]:
print_2_vezes(':-)')

:-)
:-)


In [27]:
print_2_vezes(86)

86
86


In [28]:
from math import cos, pi
print_2_vezes(cos(pi))

-1.0
-1.0


Você também pode usar uma variável como argumento:

In [29]:
nome = 'Tiago Pessoa Ferreira de Lima'
print_2_vezes(nome)

Tiago Pessoa Ferreira de Lima
Tiago Pessoa Ferreira de Lima


### **Exercícios**

1. Faça uma função que mostre o reverso de um inteiro. Por exemplo: 127 -> 721.

In [30]:
num = '123'
num[:-1]

'12'

In [31]:
def reverso_inteiro(numero):
  numero = str(numero)
  numero = numero[::-1]
  return int(numero)

reverso_inteiro(123)

321

2. Faça uma função que informe a quantidade de dígitos de um número inteiro. 


3. Faça uma função que computa $a^b$ para valores $a$ e $b$ (assuma números inteiros e não use o operador **).

## **Parâmetros Opcionais**



A função `print_2_vezes`, apresentada anteriormente, obriga que você informe um parâmetro para o argumento texto. No entanto, é possível escrever funções com argumentos opcionais. Por exemplo, a função `print_2_vezes`, apresentada abaixo, o argumento texto fará referência à String 'Tiago Lima' caso não seja informado um parâmetro na chamada da função.

In [32]:
def print_2_vezes(texto='Tiago Lima'):
  print(texto)
  print(texto)

In [33]:
print_2_vezes()

Tiago Lima
Tiago Lima


**Observação:** você já usou diversos parâmetros opcionais. Na função `print`, por exemplo, quando não especifica-se o parâmetro end, a quebra de linha '\n' é utilizada como último caracter da mensagem impressa.

## **As Variáveis e os Parâmetros são Locais**



Uma variável dentro de uma função é **local**, ou seja, ela só existe dentro da função. Exemplo:

In [34]:
def somar(operando1, operando2):
  soma = operando1 + operando2
  print(soma)

In [35]:
valor1, valor2 = 2, 3
somar(valor1, valor2)

5


Quando a função somar é encerrada, a variável soma é destruída. Se tentarmos exibí-la, recebemos uma exceção:

In [36]:
print(soma)

NameError: ignored

**Observação:** os parâmetros também são locais.

In [ ]:
print(valor1)

In [ ]:
print(valor2)

In [ ]:
print(operando1)

In [ ]:
print(operando2)

## **Passagem por Valor**

Quando se passa uma variável a uma função, o *Python* passa a referência ao objeto ao qual a variável se refere. Assim, se o valor é imutável, a função não modifica a variável chamada. Se o valor é mutável, a função pode modificar a variável chamada. Observe:

In [37]:
def modificar(x, y, z):
  x+=1
  y.append(42)
  z = [99] # nova referencia
  print('x',x)
  print('y',y)
  print('z',z)

a = 77   # variável imutável
b = [99] # variável mutável
c = [28]
modificar(a, b, c)

print('a', a)
print('b', b)
print('c', c)

x 78
y [99, 42]
z [99]
a 77
b [99, 42]
c [28]


## **Diagrama de Pilha**

Para monitorar quais variáveis podem ser usadas e onde, é uma boa ideia desenhar um diagrama da pilha. Assim como diagramas de estado, os diagramas da pilha mostram o valor de cada variável, mas também mostram a função à qual cada variável pertence.

Cada função é representada por um frame (quadro). Um frame é uma caixa com o nome de uma função junto a ele e os parâmetros e as variáveis da função dentro dele. O diagrama da pilha para o exemplo anterior é exibido na animação abaixo:

In [38]:
from IPython.display import IFrame

IFrame(src='https://pythontutor.com/iframe-embed.html#code=def%20modificar%28x,%20y,%20z%29%3A%0A%20%20x%2B%3D1%0A%20%20y.append%2842%29%0A%20%20z%20%3D%20%5B99%5D%20%23%20nova%20referencia%0A%20%20print%28\'x\',x%29%0A%20%20print%28\'y\',y%29%0A%20%20print%28\'z\',z%29%0A%0Aa%20%3D%2077%20%20%20%23%20vari%C3%A1vel%20imut%C3%A1vel%0Ab%20%3D%20%5B99%5D%20%23%20vari%C3%A1vel%20mut%C3%A1vel%0Ac%20%3D%20%5B28%5D%0Amodificar%28a,%20b,%20c%29%0A%0Aprint%28\'a\',%20a%29%0Aprint%28\'b\',%20b%29%0Aprint%28\'c\',%20c%29&codeDivHeight=400&codeDivWidth=350&cumulative=false&curInstr=0&heapPrimitives=nevernest&origin=opt-frontend.js&py=3&rawInputLstJSON=%5B%5D&textReferences=false', 
       width=800, 
       height=500)

## **Número de parâmetros variáveis**


É possível especifica um função que pode receber uma quantidade indefinida de argumentos. Esses parâmetros serão "empacotados" em uma tupla chamada args. No código abaixo, note que é preciso marcar a tupla *args* com um asterisco, para que *Python* saiba que deve alocar todos os parâmetros como elementos dela, na ordem que são passados.

In [39]:
def func_args(*args):
  print(f'tipo: {type(args)} conteúdo: {args}')
  for arg in args:
    print(f'tipo: {type(arg)} conteúdo: {arg}')

func_args(1, 'A', {'valor': 10})

tipo: <class 'tuple'> conteúdo: (1, 'A', {'valor': 10})
tipo: <class 'int'> conteúdo: 1
tipo: <class 'str'> conteúdo: A
tipo: <class 'dict'> conteúdo: {'valor': 10}


*Python* também permite que uma função seja chamada com uma quantidade indefinida de parâmetros por nome. Para isso, é preciso declarar o parâmetro *kwargs*, que irá representar um dicionário em que *Python* irá alocar todos os parâmetros passados por nome que não foram pré-definidos na chamada da função. O parâmetro *kwargs* deve ser marcado por asteriscos duplos na definição da função. Exemplo:

In [40]:
def func_kwargs(**kwargs):
    print(f'tipo: {type(kwargs)} conteúdo: {kwargs}')
    for key, value in kwargs.items():
        print(f'atributo: {key}, valor: {value}')

func_kwargs(nome='James', sobrenome='Bond', cargo='Agente 007')

tipo: <class 'dict'> conteúdo: {'nome': 'James', 'sobrenome': 'Bond', 'cargo': 'Agente 007'}
atributo: nome, valor: James
atributo: sobrenome, valor: Bond
atributo: cargo, valor: Agente 007


**Observação:** As palavras args e kwargs são usadas como convenção, ou seja, poderiam ter qualquer outro nome, por exemplo: \*tiago e **lima.

### **Exercício**

Apresente uma função com parâmetros normais e quantidade indefinida de parâmetros posicionais e de parâmetros por nome.

In [41]:
resultado = print_2_vezes(':P')

:P
:P


## **Docstrings**

Docstring é a documentação sobre o que a função faz e seus parâmetros.

In [42]:
print(print.__doc__)

print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)

Prints the values to a stream, or to sys.stdout by default.
Optional keyword arguments:
file:  a file-like object (stream); defaults to the current sys.stdout.
sep:   string inserted between values, default a space.
end:   string appended after the last value, default a newline.
flush: whether to forcibly flush the stream.


Abaixo um exemplo de como criar documentação em suas próprias funções.

In [43]:
def nome_funcao(parametro):
    '''Descrição da função e de seus parâmetros'''
    # corpo da função
    pass

In [44]:
print(nome_funcao.__doc__)

Descrição da função e de seus parâmetros


## **Funções com Resultado e Funções Nulas**

Diversas funções que já usamos, como as funções matemáticas, devolvem resultado. Outras funções, como print, executam uma ação, mas não devolvem um valor. Se você atribuir o resultado a uma variável, recebe um valor especial chamado **None**:

In [45]:
resultado = print_2_vezes(':-)')

:-)
:-)


In [46]:
print(resultado)

None


O valor **None** não é o mesmo que a string 'None'. É um valor especial que tem seu próprio tipo:

In [47]:
print(type(resultado))

<class 'NoneType'>


As funções que criamos até aqui são nulas, ou seja, não devolvem um valor. Agora vejamos uma função que não possui um comando **return**:

In [48]:
def potencia(x, y):
  return x ** y

resultado = potencia(2,3)

print(resultado)

8


Também é possível construir funções que retornam múltiplos valores em *Python*. Vejamos uma função que recebe dois inteiros e retorna o quociente e o resto da sua divisão:

In [49]:
def div(dividendo, divisor):
  return dividendo // divisor, dividendo % divisor

resultado = div(2,3)

print(resultado)

(0, 2)


In [50]:
print(type(resultado))

<class 'tuple'>


Como se pode ver, os retornos da função `div` foram impressos entre parênteses, ou seja, na forma de uma tupla. Isso significa que Python emula retornos múltiplos de funções por meio do retorno de um único valor do tipo tupla. Essa abordagem permite desconstruir o retorno em múltiplas atribuições durante a chamada. Exemplo:

In [51]:
dividendo, divisor = 2, 3
quociente, resto = div(dividendo, divisor)
print(quociente, resto)

0 2


## **Passando Funções como Argumentos**

Como mencionado acima, a definição de uma função cria uma variável com o mesmo nome, cujo valor é do tipo função. Portanto funções também podem ser passadas como parâmetros para outras funções. O código abaixo define uma função que recebe um valor `x` e uma função `oper`, que define que operação será realizada com o valor de `x`:

In [52]:
def cubo(x):
  return x ** 3

def execute_opereraca(x, operacao):
    return operacao(x)

execute_opereraca(3, cubo)

27

## **O operador lambda**

Em *Python*, é possível usar o operador **lambda** para definir funções pequenas (que ocupam apenas uma linha). Isso pode ser útil para definir funções simples que não serão usadas em outros módulos ou em várias partes diferentes do código. O operador **lambda** retorna um valor do tipo função, portanto para usar a função assim definida, pode-se atribuí-la a uma variável:

In [53]:
quadrado = lambda x: x * x

quadrado(3)

9

## ***yield***

Observe as duas célula abaixo:

In [54]:
list_compr = [x for x in range(10)]
print(list_compr)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


In [55]:
list_gener = (x for x in range(10))
print(list_gener)

<generator object <genexpr> at 0x7f1d8712ced0>


Perceba que a primeira forma retorna a lista normalmente, usando *list* comprehension de modo padrão *Python*. A segunda retorna um objeto da classe *generator*, que "gera" os valores sob demanda. Para "gerar" os valores do `list_gener` podemos utilizar:

- Função `next`: para gerar o próximo elemento;

In [56]:
print(next(list_gener))

0


- `for`: para ir gerando nas iterações do loop.

In [57]:
for i in list_gener:
  print(i)

1
2
3
4
5
6
7
8
9


A coisa começa a ficar interessante quando estamos trabalhando com funções. Imagine que estamos escrevendo uma função para retornar a lista dos quadrados dos números de `x` até `y`. No entanto, gostaríamos de receber um elemento da lista, executar operações, depois gerar outro elemento, executar operações, e assim por diante. Veja como ficaria isso:

In [58]:
def operacoes(number):
  print('faça operações com:', number)
  
def gera_quadrados_no_intervalo(x, y):
  for i in range(x, y):
    yield i*i

x, y = 10, 20
  
for numero in gera_quadrados_no_intervalo(x, y):
  operacoes(numero)

faça operações com: 100
faça operações com: 121
faça operações com: 144
faça operações com: 169
faça operações com: 196
faça operações com: 225
faça operações com: 256
faça operações com: 289
faça operações com: 324
faça operações com: 361


Perceba, o *yield* nesse código funciona mais ou menos como um *return*, entretanto, mantém salvo o contexto da função, e quando o `for` fora da função itera novamente, a função `gera_quadrados_no_intervalo` volta a ser chamada, continuando sua execução de onde parou, mantendo o contexto da sua iteração, seguindo a sequência correta do `range(x, y)`.

Vamos agora trocar o *yield* por *return* (e alterar o loop fora da função para ter uma lista). Nesse caso o resultado será sempre 100, pois o return não salva o contexto da função, executando-a "do zero" toda vez. Veja abaixo:

In [59]:
def operacoes(number):
  print('faça operações com:', number)
  
def gera_quadrados_no_intervalo(x, y):
  for i in range(x, y):
    return i*i

x, y = 10, 20
  
for _ in range(x, y):
  numero = gera_quadrados_no_intervalo(x, y)
  operacoes(numero)

faça operações com: 100
faça operações com: 100
faça operações com: 100
faça operações com: 100
faça operações com: 100
faça operações com: 100
faça operações com: 100
faça operações com: 100
faça operações com: 100
faça operações com: 100


Mas é possível simplesmente carregar todos os valores em uma lista e iterá-los normalmente no for. Qual vantagem do *yield*?

Pense o seguinte: e se essa lista tiver 1 Milhão de itens e você continuar sem saber direito quanto tempo os cálculos vão demorar, você carregaria os 1 Milhão de itens na memória e deixaria lá até os cálculos terminarem?

Vamos usar a função `getsizeof` do `sys` para medir essa diferença. Ela retorna o tamanho em *bytes* do objeto passado por parâmetro.

In [60]:
from sys import getsizeof

um_milhao = 10**6

list_compr = [x for x in range(um_milhao)]
print(getsizeof(list_compr))

list_gener = (x for x in range(um_milhao))
print(getsizeof(list_gener))

8697472
128


Veja que uma lista com o range de 1 Milhão utiliza 8697464 de *bytes*, enquanto generator utiliza apenas 128 *bytes*. Dividindo um pelo outro (8697464/128), você chega em uma razão de aproximadamente de 68000 vezes!

# **Por que funções?**

- Criar uma nova função dá a oportunidade de nomear um grupo de instruções, o que deixa o programa mais fácil de ler e depurar.
- Funções podem tornar um programa menor, eliminando o código repetitivo. Depois, se fizer alguma alteração, basta fazê-la num só lugar.
- Funções bem projetadas podem ser úteis para muitos programas. Ou seja, você pode reutilizá-la inclusive em outros programas.

# **Depuração**

Uma das habilidades mais importantes que você vai aprender é a depuração. Leia este [artigo](https://zohaib.me/debugging-in-google-collab-notebook/) e elabore um resumo a respeito do assunto.

# **Desafios**

1. Escreva uma função chamada `most_frequent` que receba uma string e exiba as letras em ordem decrescente de frequência. Encontre amostras de texto de vários idiomas diferentes e veja como a frequência das letras varia entre os idiomas. Compare seus resultados com tabelas disponíveis na [Wikipedia](http://en.wikipedia.org/wiki/Letter_frequencies).

2. Escreva uma função chamada `sed` que receba como argumentos uma string-padrão, uma string de substituição e dois nomes de arquivo; ela deve ler o primeiro arquivo e escrever o conteúdo no segundo arquivo (criando-o, se necessário). Se a string-padrão aparecer em algum lugar do arquivo, ela deve ser substituída pela string de substituição.

# **Glossário**
- Função: uma sequência nomeada de declarações que executa alguma operação. Podem receber argumentos ou produzir um resultado;
- Definição de Função: uma instrução que cria uma função nova, especificando seu nome, parâmetros e as instruções;
- Objeto da Função: um valor é criado por uma definição de função. O nome da função é uma variável que se refere a um objeto de função;
- Cabeçalho: a primeira linha de uma função;
- Corpo: a sequência de instruções dentro de uma função
- Parâmetro: um nome usado dentro de uma função para se referir ao valor passado como argumento;
- Chamada de Função: instrução que executa uma função. É composta pelo nome da função seguido de argumentos entre parênteses;
- Variável Local: variável definida dentro de uma função e por este motivo só pode ser visualizada dentro da função;
- Valor de Retorno: o resultado de uma função;
- Função com Resultado: uma função que devolve um valor;
- Função Nula: uma função que devolve None;
- None: um valor especial apresentado por funções nulas;
- Módulo: um arquivo que contém uma coleção de funções relacionadas e outras definições;
- Instrução de Importação: uma instrução que lê um módulo e cria um objeto de módulo;
- Objeto de Módulo: um valor criado por uma instrução import que oferece acesso aos valores definidos em um módulo;
- Notação de Ponto: sintaxe para chamar funções em outro módulo - o nome do módulo seguido do ponto e o nome da função;
- Fluxo de Execução: a ordem na qual as instruções são executadas.
